src/intro-to-c/debugging/gdb-intro.md

```diff
@@ -0,0 +1,369 @@
+
+In the example below, a student is working on a program but encounters a segmentation fault (segfault). They're not sure what's causing it, but thanks to GDB, they can see what's going on "under the hood" of their program and try to solve the problem.
+
+## Case Study: A Pointer Program
+
+``` c
+#include <stdio.h>
+
+void multiply_by_two(int *current, int *arr);
+
+int main(void)
+{
+    int arr[5] = {10, 20, 30, 40, 50};
+    int *current = arr;
+    int i = 0;
+
+    printf("Before\n");
+    printf("------\n");
+
+    for (i = 0; i < 5; i++)
+    {
+        printf("arr[%d] = %d\n", i, *(current++));
+    }
+	
+    current = NULL;
+    multiply_by_two(current, arr);
+
+    printf("\nAfter\n");
+    printf("-----\n");
+
+    for (i = 0; i < 5; i++)
+    {
+        printf("arr[%d] = %d\n", i, *(current++));
+    }
+
+    printf("\nAll done!\n");
+
+    return 0;
+}
+
+void multiply_by_two(int *current, int *arr)
+{
+    int i = 0;
+    current = arr;
+
+    for (i = 0; i < 5; i++)
+    {
+        *(current++) = (*current * 2);
+    }
+	
+    current = arr;
+}
+
+```
+
+Okay, everything looks alright. Time to compile.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ gcc my_file.c
+my-pc:~intro-to-c/assignments/pointers$
+```
+
+Yes, no warnings! Okay, now let me run it.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ ./a.out
+Before
+------
+arr[0] = 10
+arr[1] = 20
+arr[2] = 30
+arr[3] = 40
+arr[4] = 50
+
+After
+-----
+Segmentation fault
+```
+
+No, not a segfault! This can't be possible--my code is *perfect*! Fine, I guess I'll just have to debug it. Time to start typing up print statements... or maybe I can try that debugger I've been reading about. What is it called again? Oh, yeah, GDB.
+
+## Compiling for Debugging
+
+Alright, first I have to recompile my code with the `-g` flag.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ gcc -g my_file.c
+my-pc:~intro-to-c/assignments/pointers$
+```
+
+## Entering GDB
+
+Now I need to pass GDB the program's name, and in I go.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ gdb a.out
+GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
+Copyright (C) 2022 Free Software Foundation, Inc.
+License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
+This is free software: you are free to change and redistribute it.
+There is NO WARRANTY, to the extent permitted by law.
+Type "show copying" and "show warranty" for details.
+This GDB was configured as "x86_64-linux-gnu".
+Type "show configuration" for configuration details.
+For bug reporting instructions, please see:
+<https://www.gnu.org/software/gdb/bugs/>.
+Find the GDB manual and other documentation resources online at:
+    <http://www.gnu.org/software/gdb/documentation/>.
+
+For help, type "help".
+Type "apropos word" to search for commands related to "word"...
+Reading symbols from a.out...
+(gdb)
+```
+
```

Comment:
I might consider adding a little blurb here that sets the scene for GDB. Something like it acts like its own mini command-line environment with its own set of commands. A lot of people learning to use GDB for the first time may not be super comfortable with command line environments still.